extension/ts/objects/auth.ts

```diff
@@ -0,0 +1,141 @@
+const REDIRECT_URI = "https://www.librarything.com";
+
+const OAUTH_KEY_LOCAL_STORAGE = "_oauth2_params";
+const API_KEY_LOCAL_STORAGE = "_api_key";
+const CLIENT_ID_LOCAL_STORAGE = "_client_id";
+
+const OAUTH_SCOPES = [
+	"https://www.googleapis.com/auth/drive",
+	"https://www.googleapis.com/auth/drive.file",
+	"https://www.googleapis.com/auth/spreadsheets",
+];
+
+interface OAuthParams {
+	access_token: string;
+	expires_in: string;
+	expiry_date: string;
+	scope: string;
+	token_type: string;
+}
+
+interface APICredentials {
+	apiKey: string;
+	clientID: string;
+}
+
+const getAPICredentials = (): APICredentials => {
+	if (!isAPICredentialsSavedToLocalStorage()) {
+		const apiKey = window.prompt("Please enter the Google Sheets API key");
+		const clientID = window.prompt("Please enter the Google Sheets client ID");
+
+		localStorage.setItem(API_KEY_LOCAL_STORAGE, apiKey);
+		localStorage.setItem(CLIENT_ID_LOCAL_STORAGE, clientID);
+	}
+
+	return {apiKey: localStorage[API_KEY_LOCAL_STORAGE], clientID: localStorage[CLIENT_ID_LOCAL_STORAGE]};
+};
+
+const clearAPICredentials = () => {
+	localStorage.removeItem(API_KEY_LOCAL_STORAGE);
+	localStorage.removeItem(CLIENT_ID_LOCAL_STORAGE);
+};
+
+const isAPICredentialsSavedToLocalStorage = () =>
+	!!localStorage[API_KEY_LOCAL_STORAGE] &&
+	localStorage[API_KEY_LOCAL_STORAGE] !== "null" &&
```

Comment:
if you don't save anything or the auth flow gets into a weird state, sometimes the string value `"null"` gets saved to local storage